`ldiv!` for LU Decomposition

SBuercklin opened the pull request:
Implements `ldiv!(::LU, ::CuVecOrMat)`, addressing Chris's comment at the end of the OP for #1193 

There are some extra type parameters on one of the two `ldiv!` methods because there's a method ambiguity with a similar method in `LinearAlgebra.jl`. If there's a better way to resolve the ambiguity with one method, I'll happy make the change

Commit:
add ldiv! methods for LU

Commit:
add tests for LU ldiv!

maleadt commented on lib/cusolver/linalg.jl:
```diff
@@ -338,6 +338,18 @@ LinearAlgebra.ipiv2perm(v::CuVector{T}, maxi::Integer) where T =
 end
 
+function LinearAlgebra.ldiv!(F::LU{T,<:StridedCuMatrix{T}}, B::CuVecOrMat{T}) where {T}
+    return CUSOLVER.getrs!('N', F.factors, F.ipiv, B)
```
No need for the CUSOLVER prefixes

maleadt commented on lib/cusolver/linalg.jl:
```diff
@@ -338,6 +338,18 @@ LinearAlgebra.ipiv2perm(v::CuVector{T}, maxi::Integer) where T =
+
+# LinearAlgebra.jl defines a comparable method with these restrictions on T, so we need
+#   to define with the same type parameters to resolve method ambiguity for these cases
+function LinearAlgebra.ldiv!(
```
Formatting is different from the rest of this file.

maleadt commented on lib/cusolver/linalg.jl:
```diff
@@ -338,6 +338,18 @@ LinearAlgebra.ipiv2perm(v::CuVector{T}, maxi::Integer) where T =
+function LinearAlgebra.ldiv!(
+    F::LU{T,<:StridedCuMatrix{T}}, B::CuVecOrMat{T}
+    ) where {T <: Union{Float32, Float64, ComplexF32, ComplexF64}}
+    return CUSOLVER.getrs!('N', F.factors, F.ipiv, B)
```
module prefix

maleadt commented on test/cusolver/dense.jl:
```diff
@@ -579,6 +579,17 @@ k = 1
+            lu_cpu = lu(A)
+            lu_gpu = lu(A_d)
+            @test ldiv!(lu_cpu, B) ≈ collect(ldiv!(lu_gpu, B_d))
+
```
whitespace

maleadt reviewed (commented):
I'm not familiar with this functionality, but LGTM apart from some minor nits.

maleadt commented:
Maybe @ChrisRackauckas or @kshyatt can have a quick look.

Commit:
add method for getrs! instead of ldiv!

Commit:
fix whitespace in tests

SBuercklin commented:
> I'm not familiar with this functionality, but LGTM apart from some minor nits.

Here's the relevant `ldiv!` method in `LinearAlgebra`: [link](https://github.com/JuliaLang/julia/blob/v1.8.0-rc1/stdlib/LinearAlgebra/src/lu.jl#L404-L405). This is essentially just a wrapper over `LAPACK.getrs!`. 

Another option is to extend `LinearAlgebra.LAPACK.getrs!` with the relevant `CUSOLVER.getrs!` call, with the original `getrs!` [here](https://github.com/JuliaLang/julia/blob/v1.8.0-rc1/stdlib/LinearAlgebra/src/lapack.jl#L1005-L1021). This should also solve problems anywhere else `getrs!` is called from `LinearAlgebra`, so I just pushed that change. It also requries only writing one method, instead of two, since there is no longer any method ambiguity on `ldiv!`

ChrisRackauckas reviewed (approved):
This looks right to me.

ChrisRackauckas commented:
>Another option is to extend LinearAlgebra.LAPACK.getrs! with the relevant CUSOLVER.getrs! call, with the original getrs! [here](https://github.com/JuliaLang/julia/blob/v1.8.0-rc1/stdlib/LinearAlgebra/src/lapack.jl#L1005-L1021). This should also solve problems anywhere else getrs! is called from LinearAlgebra, so I just pushed that change. It also requries only writing one method, instead of two, since there is no longer any method ambiguity on ldiv!

Generally LinearAlgebra.LAPACK is not extended. I don't know about what happens with CUDA LAPACK-like APIs though, but generally it's just the higher level LinearAlgebra APIs that are dispatched for more generic support. So I think this is fine.

maleadt reviewed (changes requested):
> Generally LinearAlgebra.LAPACK is not extended. I don't know about what happens with CUDA LAPACK-like APIs though, but generally it's just the higher level LinearAlgebra APIs that are dispatched for more generic support. So I think this is fine.

Yup, we generally don't do this. It's true that the APIs are sometimes similar enough to make that possible, but in the general case the CUDA equivalents can't be used as a drop-in replacement (with the exception of NVBLAS being a drop-in for BLAS), so we implement the higher-level APIs. That's also more robust wrt. changes in Base Julia.

Commit:
Revert "add method for getrs! instead of ldiv!"

This reverts commit 66fe58b780ee08bf8cbaea06723bfe3229ad0c29.

Commit:
address comments on ldiv! methods

SBuercklin commented:
> Yup, we generally don't do this. It's true that the APIs are sometimes similar enough to make that possible, but in the general case the CUDA equivalents can't be used as a drop-in replacement (with the exception of NVBLAS being a drop-in for BLAS), so we implement the higher-level APIs. That's also more robust wrt. changes in Base Julia.

Thanks, that makes sense. I reverted the commit and fixed the changes from above
